ports/espressif/bindings/espnow/ESPNow.c

```diff
@@ -0,0 +1,876 @@
+        return;
+    }
+
+    espnow_header_t header;
+    header.magic = ESPNOW_MAGIC;
+    header.msg_len = msg_len;
+    header.rssi = _get_rssi_from_wifi_packet(msg);
+    header.time_ms = mp_hal_ticks_ms();
+
+    ringbuf_put_n(buf, (uint8_t *)&header, sizeof(header));
+    ringbuf_put_n(buf, mac, ESP_NOW_ETH_ALEN);
+    ringbuf_put_n(buf, msg, msg_len);
+
+    self->rx_packets++;
+}
+
+// Initialize the ESP-NOW software stack,
+// register callbacks and allocate the recv data buffers.
+static void espnow_init(espnow_obj_t *self) {
+    if (espnow_deinited(self)) {
+        self->recv_buffer = m_new_obj(ringbuf_t);
+        if (!ringbuf_alloc(self->recv_buffer, self->recv_buffer_size, true)) {
+            m_malloc_fail(self->recv_buffer_size);
+        }
+
+        if (!common_hal_wifi_radio_get_enabled(&common_hal_wifi_radio_obj)) {
+            common_hal_wifi_init(false);
+            common_hal_wifi_radio_set_enabled(&common_hal_wifi_radio_obj, true);
+        }
+
+        check_esp_err(esp_wifi_config_espnow_rate(ESP_IF_WIFI_STA, self->phy_rate));
+        check_esp_err(esp_wifi_config_espnow_rate(ESP_IF_WIFI_AP, self->phy_rate));
+
+        check_esp_err(esp_now_init());
+        check_esp_err(esp_now_register_send_cb(send_cb));
+        check_esp_err(esp_now_register_recv_cb(recv_cb));
+    }
+}
+
+// De-initialize the ESP-NOW software stack,
+// disable callbacks and deallocate the recv data buffers.
+static void espnow_deinit(espnow_obj_t *self) {
+    if (self != NULL && !espnow_deinited(self)) {
+        check_esp_err(esp_now_unregister_send_cb());
+        check_esp_err(esp_now_unregister_recv_cb());
+        check_esp_err(esp_now_deinit());
+        self->recv_buffer->buf = NULL;
+        self->recv_buffer = NULL;
+        self->peers_count = 0; // esp_now_deinit() removes all peers.
+        self->tx_packets = self->tx_responses;
+    }
+}
+
+void espnow_reset(void) {
+    espnow_deinit(_get_singleton());
+    MP_STATE_PORT(espnow_singleton) = NULL;
+}
+
+// Return C pointer to byte memory string/bytes/bytearray in obj.
+// Raise ValueError if the length does not match expected len.
+static uint8_t *_get_bytes_len(mp_obj_t obj, size_t len, mp_uint_t rw) {
+    mp_buffer_info_t bufinfo;
+    mp_get_buffer_raise(obj, &bufinfo, rw);
+    mp_arg_validate_length(bufinfo.len, len, MP_QSTR_buffer);
+    return (uint8_t *)bufinfo.buf;
+}
+
+//|     def set_pmk(self, pmk: bytes) -> None:
+//|         """Set the ESP-NOW Primary Master Key (pmk) for encrypted communications.
+//|
+//|         :param bytes pmk: The ESP-NOW Primary Master Key (length = 16 bytes)."""
+//|         ...
+STATIC mp_obj_t espnow_set_pmk(mp_obj_t self_in, mp_obj_t key) {
+    check_esp_err(esp_now_set_pmk(_get_bytes_len(key, ESP_NOW_KEY_LEN, MP_BUFFER_READ)));
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_2(espnow_set_pmk_obj, espnow_set_pmk);
+
+//|     active: bool
+//|     """Initialize or de-initialize the `ESPNow` communication protocol."""
```

Comment:
Want to switch this to `deinit()` instead and then allow the constructor to make another new object instead? That would match how mdns.MDNS now works. (Except the constructor will error on second create.)

Reply:
I have a question. For singletons, what is the reasoning for implementing them in either of the following ways:
1. A static object available as a sole instance. (e.g. `wifi.radio`, `microcontroller.watchdog` etc.).
2. A dynamic object with only one instance creation allowed at a time. (e.g. `wifi.Monitor()`, `espulp.ULP()` etc.).

Reply:
1. guarantees exactly one instance of the object, 2. allows form 0 to 1 instances

Reply:
One motivation is what @deshipu said, but I'm not sure we're consistent. If you never use `wifi.Monitor()`, then why create it (saves RAM)? But there are other things where we just made it static.

Another motivation is garbage collection. We either have to write an explicit gc routine if the object is in static storage, to have the gc marker mark any Python objects in that object, or we have to include in the `ROOT_POINTERS` list.

A third reason is whether the object needs to save state before/after the lifetime of a VM. That might be true for `wifi.radio` for the web workflow, for instance (not sure). It is true for displays.